Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideAssignedOnUpdate.hs

```diff
@@ -0,0 +1,214 @@
+                  scheduledPickupTime = fromMaybe (error "latestScheduledBooking not found") driverInfo.latestScheduledBooking
+                  scheduledPickupTimeWithGraceTime = addUTCTime (transporterConfig.graceTimeForScheduledRidePickup) scheduledPickupTime
+              if expectedEndTime > scheduledPickupTimeWithGraceTime
+                then do
+                  let cReason = "Ride is Cancelled because driver can't reach pickup of its scheduled booking on time"
+                  cancelOrReallocate ride driverId cReason False
+                  return $ Terminate "Ride is Cancelled because driver can't reach pickup of its scheduled booking on time"
+                else do
+                  let rescheduleTime = addUTCTime (transporterConfig.scheduledRideJobRescheduleTime) now
+                  return $ ReSchedule rescheduleTime
+            (_) -> do
+              let cReason = "Ride is Reallocated due to some unexpected Error in allocator"
+              cancelOrReallocate ride driverId cReason True
+              return $ Terminate "Ride is Reallocated due to some unexpected Error in allocator"
+        (_, _) -> do
+          return $ Terminate "Ride is Cancelled and Reallocated"
+
+cancelOrReallocate ::
+  ( MonadFlow m,
+    EsqDBFlow m r,
+    EncFlow m r,
+    HasHttpClientOptions r c,
+    HasShortDurationRetryCfg r c,
+    CacheFlow m r,
+    HasField "modelNamesHashMap" r (HMS.HashMap Text Text),
+    HasFlowEnv m r '["nwAddress" ::: BaseUrl],
+    HasField "s3Env" r (S3.S3Env m),
+    LT.HasLocationService m r,
+    HasFlowEnv m r '["ondcTokenHashMap" ::: HMS.HashMap KeyConfig TokenConfig],
+    HasFlowEnv m r '["internalEndPointHashMap" ::: HMS.HashMap BaseUrl BaseUrl],
+    HasFlowEnv m r '["kafkaProducerTools" ::: KafkaProducerTools],
+    EsqDBReplicaFlow m r,
+    HasField "searchRequestExpirationSeconds" r NominalDiffTime,
+    HasField "jobInfoMap" r (M.Map Text Bool),
+    HasField "maxShards" r Int,
+    HasField "schedulerSetName" r Text,
+    HasField "schedulerType" r SchedulerType,
+    Metrics.HasSendSearchRequestToDriverMetrics m r,
+    HasLongDurationRetryCfg r c,
+    HasField "singleBatchProcessingTempDelay" r NominalDiffTime,
+    TranslateFlow m r,
+    HasFlowEnv m r '["maxNotificationShards" ::: Int],
+    Redis.HedisFlow m r,
+    EventStreamFlow m r,
+    HasField "minTripDistanceForReferralCfg" r (Maybe HighPrecMeters),
+    Metrics.HasCoreMetrics r
+  ) =>
+  DRide.Ride ->
+  Id DPerson.Person ->
+  Text ->
+  Bool ->
+  m ()
+cancelOrReallocate ride driverId cReason isForceReallocation = do
+  let driverReq =
+        RideCancel.CancelRideReq
+          { reasonCode = DCR.CancellationReasonCode cReason,
+            additionalInfo = Nothing
+          }
+  (_cancellationCnt, _isGoToDisabled) <- RideCancel.cancelRideImpl RideCancel.cancelRideHandle (RideCancel.PersonRequestorId driverId) ride.id driverReq isForceReallocation
```

Comment:
We should consider the cancellationSource as ByApplication here.

Reply:
okay